### PR TITLE
Increase error counter on write-cache flush errors

### LINKED ISSUE #1818: Write cache flush errors doesn't counted in shard errors
## Expected Behavior
All  input/output errors should count as shard errors

## Current Behavior
Write cache flush errors doesn't counted in shard errors

## Steps to Reproduce (for bugs)
1. I see the input/output errors in log:
```
сен 27 14:46:06 vedi neofs-node[4612]: 2022-09-27T14:46:06.031Z        error        writecache/flush.go:161        cant flush object to blobstor        {"shard_id": "QGASQM25MvVMdJNixnsuS6", "error": "write /data/neofs/data1/7/h/6/y/qU7rTfTVSKuiKc6hbwPECSBPJzVfPJZps4hVdNj8.GGBYm5zFoGXf874N6LV4yiYhEkT7kBUzFtNo5sd2pmaT: input/output error"}
```
but they didn't counted in shard errors:
```
root@vedi:~# sudo neofs-cli --endpoint 127.0.0.1:8091 -w /etc/neofs/storage/wallet.json control shards list

Enter password >
Shard QGASQM25MvVMdJNixnsuS6:
Mode: read-write
Metabase: /srv/neofs/meta0/metabase1.db
Blobstor: /data/neofs/data1
Write-cache: /srv/neofs/meta0/write_cache1
Pilorama: /srv/neofs/meta0/pilorama1.db
Error count: 0
...
```
Log:  [write_cache_errors.zip](https://github.com/nspcc-dev/neofs-node/files/9657922/write_cache_errors.zip)

**Versions:**
```
NeoFS Storage node
Version: v0.32.0-12-gc7f85994
GoVersion: go1.18.4
```
## Your Environment
Server setup and configuration:
cloud, 4 VMs, 4 SN, 4 http qw, 4 s3 gw

Operating System and version (uname -a):
linux vedi 5.10.0-16-amd64 https://github.com/nspcc-dev/neofs-node/issues/1 SMP Debian 5.10.127-1 (2022-06-30) x86_64 GNU/Linux

### PR DESCRIPTION
**Pull request by fyrchik**
Close #1818.

### CODE REVIEW

File: pkg/local_object_storage/metabase/storage_id.go [hunk 1]
```diff
@@ -52,3 +52,41 @@ func (db *DB) storageID(tx *bbolt.Tx, addr oid.Address) ([]byte, error) {
...
+// UpdateStorageIDPrm groups the parameters of UpdateStorageID operation.
+type UpdateStorageIDPrm struct {
+	addr oid.Address
+	id   []byte
+}
+
+// UpdateStorageIDRes groups the resulting values of UpdateStorageID operation.
+type UpdateStorageIDRes struct{}
+
+// SetAddress is an UpdateStorageID option to set the object address to check.
```
Comment (carpawell) [inline pkg/local_object_storage/metabase/storage_id.go:65]: not use "option"?

File: pkg/local_object_storage/metabase/storage_id.go [hunk 2]
```diff
@@ -52,3 +52,41 @@ func (db *DB) storageID(tx *bbolt.Tx, addr oid.Address) ([]byte, error) {
...
+func (p *UpdateStorageIDPrm) SetAddress(addr oid.Address) {
+	p.addr = addr
+}
+
+// SetStorageID is an UpdateStorageID option to set the storage ID.
+func (p *UpdateStorageIDPrm) SetStorageID(id []byte) {
+	p.id = id
+}
+
+// UpdateStorageID returns storage descriptor for objects from the blobstor.
```
Comment (carpawell) [inline pkg/local_object_storage/metabase/storage_id.go:75]: wrong description?

File: pkg/local_object_storage/engine/shards.go [hunk 3]
```diff
@@ -87,6 +87,9 @@ func (e *StorageEngine) createShard(opts []shard.Option) (*shard.Shard, error) {
 		shard.WithExpiredTombstonesCallback(e.processExpiredTombstones),
 		shard.WithExpiredLocksCallback(e.processExpiredLocks),
 		shard.WithDeletedLockCallback(e.processDeletedLocks),
+		shard.WithReportErrorFunc(func(msg string, err error) {
+			e.reportShardErrorBackground(id.String(), msg, err)
```
Comment (carpawell) [inline pkg/local_object_storage/engine/shards.go:91]: Could be a bug here like with metrics: `id` is generated, but could be updated below.

File: pkg/local_object_storage/engine/engine.go [hunk 4]
```diff
@@ -33,6 +33,26 @@ type shardWrapper struct {
...
+	e.mtx.RLock()
+	sh, ok := e.shards[id]
+	e.mtx.RUnlock()
+
+	if !ok {
+		return
+	}
+
+	errCount := sh.errorCount.Inc()
+	e.log.Warn(msg, append([]zap.Field{
```
Comment (carpawell) [inline pkg/local_object_storage/engine/engine.go:49]: why `append` here?

Comment (cthulhu-rider) [reply pkg/local_object_storage/engine/engine.go:49]: Linter watches I guess.

File: pkg/local_object_storage/engine/engine.go [hunk 5]
```diff
@@ -33,6 +33,26 @@ type shardWrapper struct {
 	*shard.Shard
 }
 
+// reportShardErrorBackground increases shard error counter and logs an error.
+// It is intended to be used from background workers and
+// doesn't change shard mode because of possible deadlocks.
+func (e *StorageEngine) reportShardErrorBackground(id string, msg string, err error) {
```
Comment (cthulhu-rider) [inline pkg/local_object_storage/engine/engine.go:39]: Lets clarify that `id` is a shard ID? Naming or docs

File: pkg/local_object_storage/metabase/storage_id.go [hunk 6]
```diff
@@ -52,3 +52,41 @@ func (db *DB) storageID(tx *bbolt.Tx, addr oid.Address) ([]byte, error) {
...
+// It is put together with the object can makes get/delete operation faster.
+func (db *DB) UpdateStorageID(prm UpdateStorageIDPrm) (res UpdateStorageIDRes, err error) {
+	db.modeMtx.RLock()
+	defer db.modeMtx.RUnlock()
+
+	currEpoch := db.epochState.CurrentEpoch()
+
+	err = db.boltDB.Batch(func(tx *bbolt.Tx) error {
+		exists, err := db.exists(tx, prm.addr, currEpoch)
+		if !exists || err != nil {
```
Comment (cthulhu-rider) [inline pkg/local_object_storage/metabase/storage_id.go:85]: Can `err` be non-nil if `exists` is true?

File: pkg/local_object_storage/metabase/storage_id.go (hunk 1 above)
Comment (carpawell) [reply pkg/local_object_storage/metabase/storage_id.go:65]: just want to mention that it is a setter, not an option (functional i mean, like we sometimes do), and recheck that you write it consciously. i have nothing against such wording

**Review by carpawell — APPROVED** [review]

File: pkg/local_object_storage/shard/shard.go [hunk 7]
```diff
@@ -96,13 +96,16 @@ type cfg struct {
...
 
 func defaultCfg() *cfg {
 	return &cfg{
-		rmBatchSize: 100,
-		log:         &logger.Logger{Logger: zap.L()},
-		gcCfg:       defaultGCCfg(),
+		rmBatchSize:     100,
+		log:             &logger.Logger{Logger: zap.L()},
+		gcCfg:           defaultGCCfg(),
+		reportErrorFunc: func(string, string, error) {},
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/local_object_storage/shard/shard.go:108]: May be, this way it is simpler for all kinds of tests.

**Comment by fyrchik** [comment]
@carpawell functional options excel for object construction IMO. For functions in the hot path `Set*` approach is better, the naming reflects the difference.

**Pull request merged**